### PR TITLE
chore: support ape version 0.4

### PR DESCRIPTION
**Pull request by unparalleled-js**
### What I did

fixes: #

### Checklist

- [ ] Passes all linting checks (pre-commit and CI jobs)
- [ ] New test cases have been added and are passing
- [ ] Documentation has been updated
- [ ] PR title follows [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) standard (will be automatically included in the changelog)

### CODE REVIEW

File: ape_starknet/transactions.py [hunk 1]
```diff
@@ -342,7 +345,30 @@ def decode_logs(self, abi: Union[EventABI, ContractEvent]) -> Iterator[ContractL
...
+                for address, contract in contract_types.items()
+            }
+            for log in log_data_items:
+                contract_address = self.starknet.decode_address(log["from_address"])
+                if contract_address not in selectors:
+                    continue
+
+                for event_key in log.get("keys", []):
+                    event_abi = selectors[contract_address][event_key]
+                    yield from self.provider.network.ecosystem.decode_logs([log], event_abi)
```
Comment (ca11ab1e) [inline ape_starknet/transactions.py:371]:
```suggestion
                    yield from self.starknet.decode_logs([log], event_abi)
```

File: tests/functional/test_networks.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
+from ape.contracts import ContractInstance
+
+
+def test_multiple_networks_in_test(
```
Comment (ca11ab1e) [inline tests/functional/test_networks.py:4]: :100:

File: tests/functional/test_receipt.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    assert log.amount0 == AMOUNT_0
+    assert log.amount1 == AMOUNT_1
+    assert log.to == int(second_account.address, 16)
+
+
+def test_decode_logs_no_specify_abi(receipt, account, second_account):
+    logs = list(receipt.decode_logs())
+    assert len(logs) == 2
+    transfer_log = logs[0]
+    mint_log = logs[1]
```
Comment (ca11ab1e) [inline tests/functional/test_receipt.py:33]: Nitpick :)
```suggestion
    transfer_log, mint_log = logs
```

File: ape_starknet/transactions.py [hunk 4]
```diff
@@ -260,6 +260,15 @@ def ran_out_of_gas(self) -> bool:
     def total_fees_paid(self) -> int:
         return 0  # Overidden by Invoke receipts
 
+    def decode_logs(
+        self,
+        abi: Optional[
+            Union[List[Union[EventABI, "ContractEvent"]], Union[EventABI, "ContractEvent"]]
+        ] = None,
```
Comment (ca11ab1e) [inline ape_starknet/transactions.py:267]: That type annotation is huge, and used twice. Should we define a specific at the top of that file with something like:

```python
SomeAbis = Optional[Union[List[Union[EventABI, "ContractEvent"]], Union[EventABI, "ContractEvent"]]]
```
(Yes, the name is not perfect ^^)

File: ape_starknet/transactions.py [hunk 5]
```diff
@@ -260,6 +260,15 @@ def ran_out_of_gas(self) -> bool:
...
+    ) -> Iterator[ContractLog]:
+        # Overriden in InvocationReceipt
```
Comment (ca11ab1e) [inline ape_starknet/transactions.py:269]: Why is that method needed if it's overriden?
If the body is never executed, could we just `pass` or something like that?

Comment (unparalleled-js) [reply ape_starknet/transactions.py:269]: needed because it is abstract in the receipt api model

**Review by ca11ab1e — APPROVED** [review]

**Pull request merged**